### PR TITLE
fix when proxy tunneling failed (IOException is hidden) JDK-8173…

### LINKED ISSUE #905: IOException wrapped in InternalError not handled properly
**Describe the bug**
The call to `socket.connect(...)` at `WebSocketClient:436` can throw a `java.lang.InternalError` that is currently not handled in the enclosing try/catch as it extends `java.lang.Error` and not `java.lang.Exception`

This `InternalError` wraps a `java.lang.IOException` that contains useful information to the websocket engine to hande this case.

**To Reproduce**
1. Install a proxy, i.e. squid
2. Make it deny connections

**Example application to reproduce the issue**
Use the previously installed proxy in the `ProxyClientExample`

**Expected behavior**
An IOException handled by the underlying websocket engine

**Debug log**
```
Exception in thread "WebSocketConnectReadThread-11" java.lang.InternalError: Should not reach here
	at java.net.HttpConnectSocketImpl.doTunneling(HttpConnectSocketImpl.java:181)
	at java.net.HttpConnectSocketImpl.doTunnel(HttpConnectSocketImpl.java:168)
	at java.net.HttpConnectSocketImpl.access$200(HttpConnectSocketImpl.java:44)
	at java.net.HttpConnectSocketImpl$2.run(HttpConnectSocketImpl.java:151)
	at java.net.HttpConnectSocketImpl$2.run(HttpConnectSocketImpl.java:149)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.HttpConnectSocketImpl.privilegedDoTunnel(HttpConnectSocketImpl.java:148)
	at java.net.HttpConnectSocketImpl.connect(HttpConnectSocketImpl.java:111)
	at java.net.Socket.connect(Socket.java:589)
	at org.java_websocket.client.WebSocketClient.run(WebSocketClient.java:436)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.net.HttpConnectSocketImpl.doTunneling(HttpConnectSocketImpl.java:179)
	... 10 more
Caused by: java.io.IOException: Unable to tunnel through proxy. Proxy returns "HTTP/1.1 403 Forbidden"
	at sun.net.www.protocol.http.HttpURLConnection.doTunneling(HttpURLConnection.java:2142)
	... 15 more
```

**Environment(please complete the following information):**
 - Version used: 1.4.0
 - Java version: 1.8.0_181
 - Operating System and version: Ubuntu 16.04 LTS
 - Endpoint Name and version: "ws://echo.websocket.org"
 - Link to your project: ProxyClientExample in the examples folder

**Additional context**
This is my first time reporting an issue on a open source project.
I'm open to feedback :)

### PR DESCRIPTION
**Pull request by PearSlice**
Handle IOException when proxy auth fails

## Description
When using an authenticated proxy to connect to a Websocket and providing wrong credentials the error is hidden.
This fix allows to have a better view on the root exception.

## Related Issue
Fixes #905

## Motivation and Context
This allows to see exception
`Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 Proxy Auth Required"`
instead of having `java.lang.InternalError: Should not reach here` masking it.

## How Has This Been Tested?

By providing wrong/no credentials to our squid proxy in prod env
By providing wrong/no credentials to a local cntlm proxy in dev env

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by marci4** [comment]
Hello @PearSlice,

could you please open a normal issue as well where you can describe the problem a bit more into detail (and maybe some steps to repeat etc etc) so other people understand the context.

Thank you!

Best regards,
Marcel

**Comment by PearSlice** [comment]
Hi,

I created the issue https://github.com/TooTallNate/Java-WebSocket/issues/905 and updated the pull request accordingly.

Best Regards,

Pascal

**Comment by marci4** [comment]
@PhilipRoman if you have the time, could you maybe do a review please? Thanks

File: src/main/java/org/java_websocket/client/WebSocketClient.java [hunk 1]
```diff
@@ -452,6 +453,15 @@ public void run() {
...
 			return;
+		} catch (InternalError e) {
+			// https://bugs.openjdk.java.net/browse/JDK-8173620
+			if (e.getCause() instanceof InvocationTargetException && e.getCause().getCause() instanceof IOException) {
+				IOException cause = (IOException) e.getCause().getCause();
+				onWebsocketError(engine, cause);
+				engine.closeConnection(CloseFrame.NEVER_CONNECTED, cause.getMessage());
+				return;
+			}
+			throw e;
```
Comment (PhilipRoman) [inline src/main/java/org/java_websocket/client/WebSocketClient.java]: Why re-throw InternalError instead of returning normally?

Comment (PearSlice) [reply src/main/java/org/java_websocket/client/WebSocketClient.java]: The idea was to handle only encapsulated `IOException` by calling the `onError` of the `WebSocketClient` and closing the connection as if we received the `IOException` directly.
This way the implementation of the `WebSocketClient` can handle the error.

If it's any other kind of InternalError we should not handle it here, the same way we don't usually include catching `Error` in all `try{...} catch{...}`

What do you think?

Comment (PhilipRoman) [reply src/main/java/org/java_websocket/client/WebSocketClient.java]: You're completely right, somehow didn't notice the 'return' statement within 'if'

**Comment by marci4** [comment]
@PhilipRoman you're happy with the PR? (I'm sorry to bother you again)

**Comment by PhilipRoman** [comment]
Yes, seems like a harmless fix. I couldn't test it though.

**Pull request merged**